src/core/net/SRP_CODER.md

```diff
@@ -0,0 +1,284 @@
+# SRP Coder
+
+This document specifies SRP coder used to encode an SRP message into a compact, compressed format, reducing the message size. On the server, the received coded message can be decoded to reconstruct an exact replica of the original message.
+
+The SRP protocol uses DNS resource record format to convey information about registered services and host addresses. The use of these records and adherence to the DNS Update general format introduces overhead and can lead to larger SRP message sizes, which is undesirable for Thread network. The SRP coder aims to optimize the message, resulting in much smaller messages containing all the same information reducing overall protocol overhead.
+
+## Compact Integer Format
+
+- Works with unsigned integer of different length (`uint16`, `uint32`)
+- Number is encoded as one or more segment(s).
+- Segments are one byte (8-bit) long except for the first segment which may have fewer bits (2-7 bits).
+- First (MSB) bit in each segment is the “continuation bit” indicating whether there are more segments to follow (`1`) or if it is the last segment (`0`).
+- The remaining bits after the MSB provide the numerical `uint` bit values in big-endian order.
+
+### Examples
+
+- Numbers less than (2^7) 128 are encoded as a single byte.
+- Numbers between 128 and (2^14 - 1) 16,383 are encoded as two bytes.
+- Numbers between 16,384 up to (2^21 - 1) 2,097,151 are encoded as three bytes.
+
+## DNS Name and Label Format
+
+### Label Dispatch
+
+```
+    0   1   2   3   4   5   6   7
+  +---+---+---+---+---+---+---+---+
+  | Type  ... | ...               |
+  +---+---+---+---+---+---+---+---+
+
+```
+
+- `00 <6-bit len>`: Standard label encoding, where the lower 6 bits indicate the label's length (number of characters). This is followed by the label characters, totaling `len` in number.
+- `01 <6-bit len>`: Similar to the above, but the label begins with an underscore `'_'`. The `len` value here doesn't include the underscore itself, which isn't explicitly encoded after the dispatch byte.
```

Comment:
Is it really useful to burn this bit combination on a leading underscore when you already have another combination that captures the majority of leading-underscore names in `110`?

Reply:
If we can identify other potential use for this pattern, we can consider repurposing it. It is useful to elide extra underscores from service names (including sub-type names) that don't follow the commonly used names (by matter).

Reply:
Yeah, this makes sense. I forgot about the service type.